### PR TITLE
feat: Monitoring

### LINKED ISSUE #91: Add basic Monitoring for key areas
**What**

We need to add a set of basic metrics around key areas in the GLBC

- [x] #122
- [x] #106
- [x] #123
- [x] #124
- [x] #125
- [x] #108
- [x] #128

### PR DESCRIPTION
**Pull request by astefanutti**
This PR enables the monitoring of the GLBC:

- [x] Exposes a `/metrics` HTTP endpoint that serves Prometheus metrics
- [x] Instruments the controllers to expose reconciliation metrics
- [x] Activates controller workqueue metrics 
- [x] Enables the _standard_ process and Go runtime metrics
- [x] Instruments Route53 client to exposes AWS SDK API request metrics
- [x] Exposes basic metrics for TLS certificate issuance
- [x] Update deployment manifests 
- [x] Adds documentation

In case DNS management is delegated to external-dns, the instrumentation of the Route53 client may still be needed for health check management. Though we might want to homogenise the corresponding metrics format.

Fixes #91

### CODE REVIEW

File: cmd/kcp-glbc/main.go [hunk 1]
```diff
@@ -177,25 +191,41 @@ func main() {
 	glbcFilteredInformerFactory.Start(ctx.Done())
 	glbcFilteredInformerFactory.WaitForCacheSync(ctx.Done())
 
-	go func() {
-		ingressController.Start(ctx, numThreads)
-	}()
+	// start listening on the metrics endpoint
+	server, err := metrics.NewServer(monitoringPort)
```
Comment (maleck13) [inline cmd/kcp-glbc/main.go:195]: nit: would it make sense to call this metricsServer

File: docs/observability/monitoring.adoc [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+glbc_tls_certificate_pending_request_count{issuer="letsencryptstaging"} 0
+# HELP glbc_tls_certificate_request_errors_total GLBC TLS certificate total number of request errors
+# TYPE glbc_tls_certificate_request_errors_total counter
+glbc_tls_certificate_request_errors_total{issuer="letsencryptstaging"} 0
+# HELP glbc_tls_certificate_request_total GLBC TLS certificate total number of requests
+# TYPE glbc_tls_certificate_request_total counter
+glbc_tls_certificate_request_total{issuer="letsencryptstaging",result="failed"} 0
+glbc_tls_certificate_request_total{issuer="letsencryptstaging",result="succeeded"} 1
+# HELP glbc_tls_certificate_secret_count GLBC TLS certificate secret count
+# TYPE glbc_tls_certificate_secret_count gauge
```
[the remark this answers is not in the document]
Comment (maleck13) [reply docs/observability/monitoring.adoc:138]: I do wonder if we should have a label/dimension that is the ingress/host ?

[the remark this answers is not in the document]
Comment (astefanutti) [reply docs/observability/monitoring.adoc:138]: Yes, that would be a good idea, at least for the `glbc_tls_certificate_pending_request_count` metric. That would enables to write rules like "`by (host), for: 10m`" to alert when a request for a specific Ingress/host takes too long. For the other metrics, I wonder if that might introduce too much cardinality, and if it really adds value aside troubleshooting which may be best addressed by logging.

File: pkg/dns/aws/client.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
+/*
```
Comment (maleck13) [inline pkg/dns/aws/client.go:1]: I am keen to get these changes in as it is a big change set. But wondering if we want to stick with this client or use the library mike found instead?

File: pkg/dns/aws/client.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+				// A service error occurred
+				code = strconv.Itoa(reqErr.StatusCode())
+			} else {
+				// Generic AWS Error with Code, Message, and original error (if any)
+				code = awsErr.Code()
+			}
+		}
+	}
+	route53RequestDuration.WithLabelValues(operation, code).Observe(duration)
+	route53RequestTotal.WithLabelValues(operation, code)
```
Comment (maleck13) [inline pkg/dns/aws/client.go:51]: assume route53RequestTotal auto increments?

File: pkg/reconciler/controller.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key,
+	// to unblock other workers.
+	defer c.Queue.Done(key)
+
+	activeWorkers.WithLabelValues(c.Name).Add(1)
+	defer activeWorkers.WithLabelValues(c.Name).Add(-1)
+
+	start := time.Now()
```
Comment (maleck13) [inline pkg/reconciler/controller.go:88]: nice simple solution 👍

File: pkg/reconciler/deployment/controller.go [hunk 6]
```diff
@@ -63,80 +63,14 @@ type ControllerConfig struct {
...
-
-	// No matter what, tell the queue we're done with this key, to unblock
-	// other workers.
-	defer c.queue.Done(key)
-
-	err := c.process(ctx, key)
-	c.handleErr(err, key)
-	return true
-}
-
```
Comment (maleck13) [inline pkg/reconciler/deployment/controller.go:118]: nice to get rid of all this repeated boiler plate

**Review by maleck13 — APPROVED** [review]
The changes look good to me. Quite a bit of clean up also. I can see how the TLS metrics were difficult to add. I think we should merge once you are happy with your testing on cluster

File: cmd/kcp-glbc/main.go (hunk 1 above)
Comment (astefanutti) [reply cmd/kcp-glbc/main.go]: Right, updated.

File: docs/observability/monitoring.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (astefanutti) [reply docs/observability/monitoring.adoc:138]: I've added the `hostname` dimension to all the TLS certificates related metrics.

File: pkg/dns/aws/client.go (hunk 4 above)
Comment (astefanutti) [reply pkg/dns/aws/client.go:51]: Good catch! Fixed.

File: pkg/dns/aws/client.go (hunk 3 above)
Comment (astefanutti) [reply pkg/dns/aws/client.go:1]: Right, I think it'll make sense to migrate to the same library that external-dns relies on to instrument Route53 API calls, but I'd be inclined to do it once we do the actual migration and also cover cert-manager.

**Comment by astefanutti** [comment]
> The changes look good to me. Quite a bit of clean up also. I can see how the TLS metrics were difficult to add.

#83 should help making the TLS metrics easier to reason about.

> One other thing, wondering how KCP registers the metrics? I see we went with quite a few separate files in each package.

I haven't seen anything special in KCP yet about metrics. The idea of having one `metrics` file in each package, is to encapsulate the metrics observations, and also help enforcing "symmetry", especially for gauges. In the case of the TLS certificate duration, I had to break that pattern, which is symptomatic of a dilution of control / responsibility.

File: config/prometheus/pod_monitor.yaml [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  labels:
+    app.kubernetes.io/name: kcp-glbc
+    app.kubernetes.io/component: controller-manager
+spec:
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: kcp-glbc
+      app.kubernetes.io/component: controller-manager
+  podMetricsEndpoints:
+    - port: metrics
```
Comment (mikenairn) [inline config/prometheus/pod_monitor.yaml:14]: Cool, wondered about this. What ns does this get created in?

**Comment by astefanutti** [comment]
I've successfully tested the PR, and I think it's ready to be merged.

Here are the follow-up items:
- #83 -> Improve metrics observation
- #92
- Basic alerting
- Update DNSRecord and Route53 client instrumentation once we migrate to using external-dns
- Add e2e tests

**Comment by maleck13** [comment]
for alerting, it will be unique per users requirements, so I think that is something we wont dictate / supply as part of the repo

**Comment by astefanutti** [comment]
> for alerting, it will be unique per users requirements, so I think that is something we wont dictate / supply as part of the repo

Right, I was thinking providing a PrometheusRule resource template, that can be used to quick start alerting, but meant to be further tailored to user requirements.

**Pull request merged**